### ISSUE TITLE
Can't run a container when mounting /dev:/dev

### ISSUE THREAD

**Issue by eadamsintel** [kata-containers/kata-containers#692]
# Description of problem

When trying to run the following commands I get errors when trying to mount `/dev:/dev` with containerd and Kata 2.0 alpha 3. Another user also saw errors when doing this with `/dev:/dev` but `/dev/sgx:/dev/sgx` did work for them on their newer version of Kata recently built from master.  However, when trying with `/dev/qat_adf_ctl:/dev/qat_adf_ctl` it failed for me on Alpha 3.  The following works just fine with Kata v1.11.3.

Here are some commands showing that it works with just a simple file mount but not with `/dev` or `/dev/qat_adf_ctl`

```
eadams@otc-tme-wf-02~ $ sudo ctr run --runtime io.containerd.run.kata-qemu.v2 -t --rm --privileged --mount type=bind,src=/home/eadams,dst=/eadams,options=rbind:rw  docker.io/library/ubuntu:latest bash
root@clr-9f6a1f4ede17491ba755ec6ee153e4db:/# exit
exit
eadams@otc-tme-wf-02~ $ sudo ctr run --runtime io.containerd.run.kata-qemu.v2 -t --rm --privileged --mount type=bind,src=/dev,dst=/dev,options=rbind:rw  docker.io/library/ubuntu:latest bash
ctr: Error Code: 'Error(Io(Os { code: 17, kind: AlreadyExists, message: "File exists" }), State { next_error: None, backtrace: InternalBacktrace { backtrace: Some(   0: error_chain::backtrace::imp::InternalBacktrace::new
   1: <error_chain::State as core::default::Default>::default
   2: rustjail::mount::init_rootfs
   3: rustjail::container::do_init_child
   4: rustjail::container::init_child
   5: kata_agent::main
   6: std::rt::lang_start::{{closure}}
   7: std::rt::lang_start_internal::{{closure}}
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/rt.rs:52
      std::panicking::try::do_call
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panicking.rs:297
      std::panicking::try
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panicking.rs:274
      std::panic::catch_unwind
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panic.rs:394
      std::rt::lang_start_internal
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/rt.rs:51
   8: main
) } })': unknown
eadams@otc-tme-wf-02~ $ sudo ctr run --runtime io.containerd.run.kata-qemu.v2 -t --rm --privileged --mount type=bind,src=/dev/qat_adf_ctl,dst=/dev/qat_adf_ctl,options=rbind:rw  docker.io/library/ubuntu:latest bash
ctr: Error Code: 'Error(Io(Os { code: 2, kind: NotFound, message: "No such file or directory" }), State { next_error: None, backtrace: InternalBacktrace { backtrace: Some(   0: error_chain::backtrace::imp::InternalBacktrace::new
   1: <error_chain::State as core::default::Default>::default
   2: rustjail::mount::mount_from
   3: rustjail::mount::init_rootfs
   4: rustjail::container::do_init_child
   5: rustjail::container::init_child
   6: kata_agent::main
   7: std::rt::lang_start::{{closure}}
   8: std::rt::lang_start_internal::{{closure}}
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/rt.rs:52
      std::panicking::try::do_call
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panicking.rs:297
      std::panicking::try
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panicking.rs:274
      std::panic::catch_unwind
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/panic.rs:394
      std::rt::lang_start_internal
             at /rustc/5c1f21c3b82297671ad3ae1e8c942d2ca92e84f2/src/libstd/rt.rs:51
   9: main
) } })': unknown

```

# Expected result

The container should start.

# Actual result

I get the error messages above.

**Comment by eadamsintel** [comment]
I tried this with the latest RC1 and the error is slightly different but still happens.

```
$ sudo ctr run --runtime io.containerd.run.kata-qemu.v2 -t --rm --privileged --mount type=bind,src=/dev,dst=/dev,options=rbind:rw  docker.io/library/ubuntu:latest bash
ctr: File exists (os error 17): unknown

$ sudo ctr version
Client:
  Version:  1.3.4
  Revision:

Server:
  Version:  1.3.4
  Revision:
  UUID: 60c23008-b627-4b00-9db1-c93bea412aa1

```

**Comment by ariel-adam** [comment]
@eadamsintel any chance if you could test it again since we released an official 2.0 since then?

**Comment by eadamsintel** [comment]
@ariel-adam I was on vacation and just saw your message. I tried to cobble together a release building from master yesterday but had issues building a rootfs, so I used an initrfs. For some reason I couldn't get `--privileged `to work.  I'll try later today with the 2.0 release from October but there were only a few weeks between the RC I tested and the final 2.0 release in October so I am not too hopeful but I can try.

**Comment by eadamsintel** [comment]
@ariel-adam I tried the 2.0.0 release but am getting an ENOENT error when I try to just run privileged.
Below is the privileged error I am seeing. I'll try this again when the next tarball is released.
```
###This works
eadams@otc-tme-wf-02/opt/kata/bin $ sudo ctr run --runtime "io.containerd.kata-qemu.v2" -t --rm  --mount type=bind,src=/home/eadams,dst=/eadams,options=rbind:rw  docker.io/library/ubuntu:latest bash
root@clr-f0e9dc9b6b6d4efaab24c3ae31036c7c:/# exit
exit

###This fails with privileged
eadams@otc-tme-wf-02/opt/kata/bin $ sudo ctr run --runtime "io.containerd.kata-qemu.v2" -t --rm --privileged  --mount type=bind,src=/home/eadams,dst=/eadams,options=rbind:rw  docker.io/library/ubuntu:latest bash
ctr: ENOENT: No such file or directory: unknown
```
I tried both with and without the --privileged option on the 2.0.0 release
```
eadams@otc-tme-wf-02/opt/kata/bin $ sudo ctr run --runtime io.containerd.run.kata-qemu.v2 -t --rm --privileged --mount type=bind,src=/dev,dst=/dev,options=rbind:rw  docker.io/library/ubuntu:latest bash
ctr: File exists (os error 17): unknown

eadams@otc-tme-wf-02/opt/kata/bin $ sudo ctr run --runtime io.containerd.run.kata-qemu.v2 -t --rm --mount type=bind,src=/dev,dst=/dev,options=rbind:rw  docker.io/library/ubuntu:latest bash ctr: File exists (os error 17): unknown
```

**Comment by eadamsintel** [comment]
I tried this with the 2.0 release from October, and it was the same.

```
###This works
eadams@otc-tme-wf-02/usr/local/bin $ sudo ctr run --runtime "io.containerd.kata.v2" -t --rm  --mount type=bind,src=/home/eadams,dst=/eadams,options=rbind:rw  docker.io/library/ubuntu:latest bash
root@clr-d1289353a1dc4de4bfb2af7214e98b88:/# exit
exit

###This is an issue with 2.0 with privileged
eadams@otc-tme-wf-02/usr/local/bin $ sudo ctr run --runtime "io.containerd.kata.v2" -t --rm --privileged --mount type=bind,src=/home/eadams,dst=/eadams,options=rbind:rw  docker.io/library/ubuntu:latest bash
ctr: ENOENT: No such file or directory: unknown

###The /dev mount fails
eadams@otc-tme-wf-02/usr/local/bin $ sudo ctr run --runtime "io.containerd.kata.v2" -t --rm --mount type=bind,src=/dev,dst=/dev,options=rbind:rw  docker.io/library/ubuntu:latest bash
ctr: File exists (os error 17): unknown

#The /dev mount also fails with privileged
eadams@otc-tme-wf-02/usr/local/bin $ sudo ctr run --runtime "io.containerd.kata.v2" -t --rm --privileged --mount type=bind,src=/dev,dst=/dev,options=rbind:rw  docker.io/library/ubuntu:latest bash
ctr: File exists (os error 17): unknown
```

**Comment by lifupan** [comment]
> privileged

Hi @eadamsintel 

For the privileged kata containers, you can take a look on  https://github.com/kata-containers/documentation/blob/master/how-to/privileged.md#containerd-and-cri and add an option of "privileged_without_host_devices = true" to containerd cri plugin.

**Comment by wenhuizhang** [comment]
> @eadamsintel any chance if you could test it again since we released an official 2.0 since then?

still the same here 
```
root@n192-191-015:~# ctr run --runtime io.containerd.run.kata.v2 -t --rm --privileged --mount type=bind,src=/dev,dst=/dev,options=rbind:rw   docker.io/library/busybox:latest hello10 sh
ctr: failed to create shim task: Conflicting device updates for /dev/loop7: unknown
```
